pkg/remoteimageasync/puller.go

```diff
@@ -0,0 +1,70 @@
+package remoteimageasync
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/warm-metal/container-image-csi-driver/pkg/metrics"
+	"k8s.io/klog/v2"
+)
+
+// sessionChan and completedChan both closed here
+func RunPullerLoop(
+	ctx context.Context,
+	sessionChan chan *PullSession,
+	completedChan chan string,
+) {
+	go func() {
+		<-ctx.Done()
+		close(sessionChan) // only close this once
+	}()
+	go func() {
+		wg := sync.WaitGroup{}
+		for {
+			ses, ok := <-sessionChan // ctx not observed for shut down, this sleep breaks when sessionChan is closed
+			if !ok {                 // sessionChan closed, shut down loop
+				wg.Wait()            // wait for running goroutines (which observe cancellation)
+				close(completedChan) // only close this once, after completions submitted (wg.Wait() returns)
+				return
+			}
+			wg.Add(1) // increment before goroutine starts execution to avoid race condition
+			go func() {
+				defer wg.Done() // announce done upon completion, various ways to wake this goroutine are in place
+				klog.V(2).Infof("%s.RunPullerLoop(): asked to pull image %s with timeout %v\n",
+					prefix, ses.image, ses.timeout)
+				ctxCombined, cancelDontCare := context.WithTimeout(ctx, ses.timeout) // combine timeout and shut down signal into one
+				defer cancelDontCare()                                               // IF we exit, this no longer matters. calling to satisfy linter.
+				pullStart := time.Now()
+				pullErr := ses.puller.Pull(ctxCombined) //NOTE: relying existing tests or history to verify behavior, asyncPull just wraps it
+				// update fields on session before declaring done
+				select {
+				case <-ctx.Done(): // shutting down
+					ses.isComplete = false
+					ses.isTimedOut = false
```

Comment:
Do we need `isComplete` and `isTimedOut` ?

Reply:
per our discussion earlier today, i have now removed `isComplete`. i left `isTimeout` since i felt that trying to pack the error type into the error message in a rigid way was becoming bloated and made the code harder to read and hinder debugging as the structure of logging would be changed.